### PR TITLE
RFC: Disk quota module

### PR DESCRIPTION
**Pull request by NorfairKing**
###### Description of changes

This adds a disk quota module, as proposed in https://github.com/NixOS/nixpkgs/issues/157603
and implemented first in https://github.com/NorfairKing/nix-quota-module.
I'm open to any feedback or refactoring suggestions to make this module fit into nixpkgs in the form that the maintainers would want.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NorfairKing** [comment]
CC @karantan and @zupo

**Comment by NorfairKing** [comment]
> not sure it's a good idea to split quota and fileSystems like that. 

I thought about that too but figured it might be easier to discuss like this..

File: nixos/tests/quotas.nix [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+
+    exit, out = machine.systemctl("cat set-disk-quotas.service")
+    print(out)
+    exit, out = machine.systemctl("start set-disk-quotas.service")
+    print(out)
+    if exit != 0:
+      out = machine.succeed("journalctl -u set-disk-quotas.service")
+      print(out)
+      exit(1)
```
Comment (CameronNemo) [inline nixos/tests/quotas.nix:63]: not sure what the nixos custom is, but the python docs say this about the exit builtin (and similar builtins):

> They are useful for the interactive interpreter shell and should not be used in programs.

https://docs.python.org/3/library/constants.html

Comment (CameronNemo) [reply nixos/tests/quotas.nix:63]: Seems like only `nixos/tests/matrix-appservice-irc.nix` uses the exit builtin, and only a couple tests use sys.exit. I think mose use `assert False, "message"` or similar.

Comment (NorfairKing) [reply nixos/tests/quotas.nix]: @CameronNemo Good feedback, I simplified the test.

**Review by pennae — COMMENTED** [review]
having four options that are actually a 2x2 matrix suggests there might be a better way to organise this. `quota.«fs».limits.{blocks,inodes}.{soft,hard}` might be better.

for reasons of quotas actually being tied to filesystems and not to directories that may no longer correspond to a filesystem if a mountpoint changes, quotas *really* shouldn't be separated from filesystems like this. it doesn't make review much easier, but it does introduce a ton of failure points.

can't really say whether adding a generic quota module is even a good idea. xfs for example has more quota options available (and xfs project quotas already exist, in `programs` of all places).

**Comment by karantan** [comment]
> xfs for example has more quota options available (and xfs project quotas already exist, in programs of all places).

good point. I didn't know about this project.

File: nixos/modules/tasks/filesystems/quotas.nix [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        path = [ pkgs.quota ];
+        wantedBy = [ "local-fs.target" ];
+        serviceConfig = {
+          Type = "oneshot";
+          # First turn off the quotas to safely run `quotacheck`.
+          # See https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/storage_administration_guide/disk-quotas-keep-accurate
+          ExecStart = "${pkgs.writeShellScript "set-disk-quotas" ''
+            quotaoff --all --user --group --verbose
+            quotacheck --all --user --no-remount --verbose
+            quotaon --all --user --group --verbose
```
[the remark this answers is not in the document]
Comment (NorfairKing) [reply nixos/modules/tasks/filesystems/quotas.nix:109]: Not sure, but I'm guessing this is super fast.

**Comment by NorfairKing** [comment]
@pennae I agree with all of your comments. I'm happy to refactor but it looks like we might not want this module at all or maybe we do but somewhere else? It's not really clear to me how I should proceed..
I'd like to hear @nh2 's opinion on this too. He's got a lot of experience with "how could this go wrong?" in practice.

**Comment by nh2** [comment]
I'm not experienced with quotas.

The arguments above make sense to me, given that some file systems have more and different quota options than others.

I'm not even sure where to find a good resources that describes how quotas _really work_. Where are the per-user quota limits stored, and where is the bookkeeping done? How/when does the kernel (?) check them, is it it on write syscalls?

File: nixos/tests/quotas.nix [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import ./make-test-python.nix ({pkgs, ... }:
+{ 
+  name = "quotas";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ zupo ];
```
[the remark this answers is not in the document]
Comment (NorfairKing) [reply nixos/tests/quotas.nix:5]: No, @zupo will really be the maintainer.

[the remark this answers is not in the document]
Comment (zupo) [reply nixos/tests/quotas.nix:5]: Huh, where did I sign up for this? :)

It's not realistic that I'll be able to maintain it. Maybe @karantan?

[the remark this answers is not in the document]
Comment (NorfairKing) [reply nixos/tests/quotas.nix:5]: Ah whoops, I just reused my assumption from the last thing I upstreamed. Yes @karantan should work in that case.

**Comment by NorfairKing** [comment]
Having thought about this a bit more,  I'm happy to close this PR for now, until we've had a bit more experience with quotas internally.
This topic is deeper than expected so we'll want to think it through further.
For example:
* I have no clue where and quotas are enforced
* I have no idea about the performance implications of quotas. If the enforcement happens during `write` syscalls then I can imagine there are performance implications.
* I have no idea how quotas interact with network filesystems or encrypted volumes.
* I have no idea how quotas and their setup interact with the mount order of filesystems.

**Pull request closed without merging**